subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -670,6 +678,49 @@ private void captureBuildOperationResult(BuildOperationContext context, Resolver
         });
     }
 
+    @Override
+    public ConfigurationInternal getConsistentResolutionSource() {
+        return consistentResolutionSource;
+    }
+
+    private void maybeConfigureConsistentResolution() {
+        if (consistentResolutionSource != null) {
+            assertThatConsistentResolutionIsPropertyConfigured();
+            consistentResolutionSource.getIncoming().getResolutionResult().allComponents(this::registerConsistentResolutionConstraint);
+        }
+    }
+
+    private void assertThatConsistentResolutionIsPropertyConfigured() {
+        if (!consistentResolutionSource.isCanBeResolved()) {
+            throw new InvalidUserCodeException("You can't use " + consistentResolutionSource + " as a consistent resolution source for " + this + " because it isn't a resolvable configuration.");
+        }
+        assertNoDependencyResolutionConsistencyCycle();
+    }
+
+    private void assertNoDependencyResolutionConsistencyCycle() {
+        Set<ConfigurationInternal> sources = Sets.newLinkedHashSet();
+        ConfigurationInternal src = this;
+        while (src != null) {
+            if (!sources.add(src)) {
+                String cycle = sources.stream().map(Configuration::getName).collect(Collectors.joining(" -> ")) + " -> " + getName();
+                throw new InvalidUserDataException("Cycle detected in consistent resolution sources: " + cycle);
+            }
+            src = src.getConsistentResolutionSource();
+        }
+    }
+
+    private void registerConsistentResolutionConstraint(ResolvedComponentResult result) {
+        if (result.getId() instanceof ModuleComponentIdentifier) {
+            ModuleVersionIdentifier moduleVersion = result.getModuleVersion();
+            DefaultDependencyConstraint constraint = DefaultDependencyConstraint.of(
+                moduleVersion.getGroup(),
+                moduleVersion.getName(),
+                vc -> vc.strictly(moduleVersion.getVersion()));
```

Comment:
why is this a lambda? I wonder if this is more readable if the factory method is actually a `strict` constraint factory so this would become:

```
DefaultDependencyConstraint.strictConstraint(
                moduleVersion.getGroup(),
                moduleVersion.getName(),
                moduleVersion.getVersion());
```

WDYT?

Reply:
Good idea, done